File: Problems Solutions/104. Maximum Depth of Binary Tree.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/

from utils import list_to_order_btree

class Solution:
    def maxDepth(self, root) -> int:
        if root is None:
            return 0
        
        depth = 1
        nodes = [(root, 1)]

        while nodes:
            node, lvl = nodes.pop()

            if node is None:
                continue

            depth = max(depth, lvl)
            nodes.append((node.left, lvl + 1))
            nodes.append((node.right, lvl + 1))

        return depth

tests_data = [
    ([3,9,20,None,None,15,7], 3),
    ([1,None,2], 2),
]
obj = Solution()

for data in tests_data:
    expected = data[-1]
    tree = list_to_order_btree(data[0])
    assert expected == obj.maxDepth(tree)

File: Problems Solutions/go/104. Maximum Depth of Binary Tree.go
package main

 type TreeNode struct {
    Val int
    Left *TreeNode
    Right * TreeNode
}

 type NodeLvlItem struct {
    node *TreeNode
    lvl int
}

func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }

    depth := 1
    var nodes []NodeLvlItem
    nodes = append(nodes, NodeLvlItem{root, 1})

    for len(nodes) > 0 {
        nodeLvlItem := nodes[len(nodes) - 1]
        node := nodeLvlItem.node
        lvl := nodeLvlItem.lvl

        nodes = nodes[:len(nodes) - 1]

        if lvl > depth {
            depth = lvl
        }

        if node.Left != nil {
            nodes = append(nodes, NodeLvlItem{node.Left, lvl + 1})
        }
        
        if node.Right != nil {
            nodes = append(nodes, NodeLvlItem{node.Right, lvl + 1})
        }
    }

    return depth
}